projects/power_distribution/test/test_fan_ctrl.c

```diff
@@ -0,0 +1,219 @@
+    .sda = TEST_CONFIG_PIN_I2C_SDA,
+  };
+  i2c_init(TEST_I2C_PORT, &i2c_settings);
+  gpio_it_init();
+}
+
+void teardown_test(void) {}
+
+void test_front_pd_fan_ctrl_init(void) {
+  TEST_ASSERT_NOT_OK(pd_fan_ctrl_init(NULL, true));
+  TEST_ASSERT_OK(pd_fan_ctrl_init(&s_fan_settings, true));
+  TEST_ASSERT_OK(pd_fan_ctrl_init(&s_fan_settings, false));
+}
+
+// Test Gpio interrupt on rear smbalert pin triggered
+void test_fan_err_rear(void) {
+  prv_initialize_can(SYSTEM_CAN_DEVICE_POWER_DISTRIBUTION_REAR);
+  TEST_ASSERT_OK(pd_fan_ctrl_init(&s_fan_settings, false));
+  gpio_it_trigger_interrupt(&(GpioAddress)REAR_PIN_SMBALERT);
+  can_register_rx_handler(SYSTEM_CAN_MESSAGE_REAR_FAN_FAULT, prv_can_fan_ctrl_rx_handler, NULL);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  TEST_ASSERT_TRUE((s_fan_ctrl_msg >> 8 & OVERVOLT_FLAGS) == OVERVOLT_FLAGS);
+  TEST_ASSERT_TRUE((s_fan_ctrl_msg & FAN_ERR_FLAGS) == FAN_ERR_FLAGS);
+  s_fan_ctrl_msg = 0;
+}
+
+// Test Gpio interrupt on front smbalert pin triggered
+void test_fan_err_front(void) {
+  prv_initialize_can(SYSTEM_CAN_DEVICE_POWER_DISTRIBUTION_FRONT);
+  TEST_ASSERT_OK(pd_fan_ctrl_init(&s_fan_settings, true));
+  gpio_it_trigger_interrupt(&(GpioAddress)FRONT_PIN_SMBALERT);
+  can_register_rx_handler(SYSTEM_CAN_MESSAGE_FRONT_FAN_FAULT, prv_can_fan_ctrl_rx_handler, NULL);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  TEST_ASSERT_TRUE((s_fan_ctrl_msg >> 8 & OVERVOLT_FLAGS) == OVERVOLT_FLAGS);
+  TEST_ASSERT_TRUE((s_fan_ctrl_msg & FAN_ERR_FLAGS) == FAN_ERR_FLAGS);
+  s_fan_ctrl_msg = 0;
+}
+
+void test_rear_pd_fan_ctrl_temp(void) {
+  // Check overtemp, undertemp, and 50% temp values transmit correct
+  // values and execute properly
+
+  // set adc to return 50% speed temp
+  adc_ret_val = FAN_50_PERC_VOLTAGE;
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, pd_fan_ctrl_init(&s_fan_settings, false));
+  delay_ms(REAR_FAN_CONTROL_REFRESH_PERIOD_MILLIS + 10);
+  TEST_ASSERT_EQUAL(FAN_50_PERC_I2C_WRITE, i2c_buf1[1]);
+  TEST_ASSERT_EQUAL(FAN_50_PERC_I2C_WRITE, i2c_buf2[1]);
+
+  // set adc to return undertemp
+  adc_ret_val = FAN_UNDERTEMP_VOLTAGE;
+  delay_ms(REAR_FAN_CONTROL_REFRESH_PERIOD_MILLIS);
+  TEST_ASSERT_EQUAL(FAN_MIN_I2C_WRITE, i2c_buf1[1]);
+  TEST_ASSERT_EQUAL(FAN_MIN_I2C_WRITE, i2c_buf2[1]);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  // Test adc returns an overtemp value - CAN message with overtemp values transmitted
+  Event e = { 0 };
+  adc_ret_val = FAN_OVERTEMP_VOLTAGE;
+  prv_initialize_can(SYSTEM_CAN_DEVICE_POWER_DISTRIBUTION_REAR);
+  delay_ms(REAR_FAN_CONTROL_REFRESH_PERIOD_MILLIS);
+  TEST_ASSERT_EQUAL(FAN_MAX_I2C_WRITE, i2c_buf1[1]);
+  TEST_ASSERT_EQUAL(FAN_MAX_I2C_WRITE, i2c_buf2[1]);
+  can_register_rx_handler(SYSTEM_CAN_MESSAGE_REAR_FAN_FAULT, prv_can_fan_ctrl_rx_handler, NULL);
+  MS_TEST_HELPER_CAN_TX(TEST_CAN_EVENT_TX);
+  while (event_process(&e) != STATUS_CODE_OK) {
+  }
+  can_process_event(&e);
```

Comment:
Was `MS_TEST_HELPER_CAN_TX_RX` not working? In any case you can do `MS_TEST_HELPER_AWAIT_EVENT(e)` instead of the while loop.